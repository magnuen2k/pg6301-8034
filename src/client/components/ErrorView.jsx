import React from "react";

export const ErrorView = ({ error }) => {
  return <div>An error has occurred: {error.toString()}</div>;
};
