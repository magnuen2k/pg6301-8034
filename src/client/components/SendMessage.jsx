import React, { useEffect, useState } from "react";
import { Input } from "./Input";

export const SendMessage = ({ username, onSendMessage, messageApi }) => {
  const [users, setUsers] = useState();
  const [recipient, setRecipient] = useState("");
  const [formData, setFormData] = useState({
    to: [],
    message: "",
    from: username,
  });

  useEffect(() => {
    messageApi.getUsersToMessage().then((res) => setUsers(res));
  }, []);

  const handleChange = (e) => {
    setFormData({
      ...formData,
      [e.target.name]: e.target.value,
    });
  };

  const handleAddRecipient = (e) => {
    let newTo = formData.to;
    newTo.push(recipient);
    e.preventDefault();
    setFormData({ ...formData, to: newTo });
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    onSendMessage(formData);
  };

  return (
    <div className="sendMessageWrapper">
      <div className="recipients">Sender: {formData && formData.from}</div>
      <div className="recipients">
        Recipients: {formData && formData.to.map((u) => u + " ")}
      </div>
      <hr />
      <form className="addRecipient" onSubmit={handleAddRecipient}>
        <p>Chose who you want to send a message to</p>
        <label>
          Pick recipients
          <select
            value={recipient}
            name="recipients"
            onChange={(e) => setRecipient(e.target.value)}
          >
            <option value="" disabled>
              --------
            </option>
            {users &&
              users.map((u) => (
                <option key={u} value={u}>
                  {u}
                </option>
              ))}
          </select>
          <Input
            name="recipient"
            label="Type recipient"
            handleChange={(e) => setRecipient(e.target.value)}
          />
        </label>
        <button className="recipientBtn" type="submit">
          +
        </button>
        <hr />
      </form>
      <form className="sendMessage" onSubmit={handleSubmit}>
        <label>
          Your message: <textarea name="message" onChange={handleChange} />
        </label>
        <button type="submit">Send message</button>
      </form>
    </div>
  );
};
