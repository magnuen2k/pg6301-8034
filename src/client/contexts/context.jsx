import React, { createContext } from "react";

export const UserContext = createContext(null);
export const WsContext = createContext(null);
