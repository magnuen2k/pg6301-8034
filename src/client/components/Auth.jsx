import React, { useContext, useState } from "react";
import { Link, useHistory } from "react-router-dom";
import { UserContext, WsContext } from "../contexts/context";
import { Input } from "./Input";

export const Auth = ({ authApi }) => {
  const [isSignup, setIsSignup] = useState(false);
  const [formData, setFormData] = useState({
    username: "",
    password: "",
  });
  const [errorCode, setErrorCode] = useState();

  const { user, setUser } = useContext(UserContext);
  const { ws } = useContext(WsContext);

  const history = useHistory();

  const handleSubmit = async (e) => {
    e.preventDefault();
    setErrorCode(undefined);
    if (isSignup) {
      // Handle signup
      await doSignUp();
    } else {
      await doSignIn();
    }
  };

  const doSignUp = async () => {
    try {
      const res = await authApi.signUp(formData);
      if (res) {
        // Set user
        setUser(res);
        ws.send(user.username);
        // Redirect user back to home page
        history.push("/");
      }
    } catch (e) {
      setErrorCode(e.response);
    }
  };

  const doSignIn = async () => {
    try {
      const res = await authApi.signIn(formData);
      if (res) {
        // Set user
        setUser(res);
        console.log(res.username);
        ws.send(res.username);
        // Redirect user back to home page
        history.push("/");
      }
    } catch (e) {
      console.log(e);
      setErrorCode(e.response);
    }
  };

  const handleChange = (e) => {
    setFormData({
      ...formData,
      [e.target.name]: e.target.value,
    });
  };

  const switchMode = () => {
    setIsSignup((prevIsSignup) => !prevIsSignup);
    // Find a way to clear input fields
  };

  const googleLogin = () => {
    window.open("http://localhost:3000/api/auth/google", "_self");
  };

  return (
    <div className="container">
      {user ? (
        <div>
          <p>You are already logged in</p>
          <Link to={"/"}>Go back to home page</Link>
        </div>
      ) : (
        <div>
          <form onSubmit={handleSubmit}>
            <Input
              name="username"
              label="Username"
              handleChange={handleChange}
            />
            <Input
              name="password"
              label="Password"
              handleChange={handleChange}
              type="password"
            />
            {isSignup && (
              <>
                <Input
                  name="firstName"
                  label="First Name"
                  handleChange={handleChange}
                />
                <Input
                  name="lastName"
                  label="Last Name"
                  handleChange={handleChange}
                />
                <Input
                  name="email"
                  label="Email"
                  handleChange={handleChange}
                  type="email"
                />
              </>
            )}
            <button id="submitBtn" type="submit">
              {isSignup ? "Sign Up" : "Sign In"}
            </button>
          </form>
          <button id="switch" onClick={switchMode}>
            {isSignup
              ? "Already have an account? Sign In"
              : "Dont have an account yet? Sign Up"}
          </button>
        </div>
      )}
      {errorCode && (
        <div>
          {errorCode === 400
            ? "Fill out the form"
            : "Wrong username or password"}
        </div>
      )}
      <div className="googleLogin" onClick={googleLogin}>
        <p>
          Click here to login with <strong>GOOGLE</strong>
        </p>
      </div>
    </div>
  );
};
