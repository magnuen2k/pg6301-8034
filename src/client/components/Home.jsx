import React, { useContext, useEffect, useRef, useState } from "react";
import { UserContext, WsContext } from "../contexts/context";
import { useWs } from "../hooks/useWs";

export const Home = () => {
  const { user } = useContext(UserContext);
  const { ws } = useContext(WsContext);

  if (!user) {
    return <div className="container">Sign in to use out service!</div>;
  }

  return (
    <div className="container">
      <h1>Welcome back, {user.firstName}</h1>
      <p>App description</p>
    </div>
  );
};
