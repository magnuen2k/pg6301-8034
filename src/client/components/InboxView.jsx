import React, { useContext } from "react";
import { UserContext } from "../contexts/context";

export const InboxView = ({ data, onReply, onReplyAll, onDelete }) => {
  const { user } = useContext(UserContext);
  return (
    <div className="container">
      <div>
        {data &&
          data
            .slice(0)
            .reverse()
            .map((msg) => (
              <div className="message" key={msg.mid}>
                <div>
                  <div>
                    From: {msg.from} ({msg.time}) | Message: {msg.content}
                  </div>
                  <div>
                    To: {user.username}{" "}
                    {msg.to && msg.to.map((u) => u.to + " ")}
                  </div>
                </div>
                {onReply ? (
                  <div>
                    <button onClick={() => onReply(msg.mid, msg.from)}>
                      Reply
                    </button>
                    <button
                      onClick={() => {
                        console.log(msg.to.map((u) => u.to));
                        onReply(
                          msg.mid,
                          msg.from,
                          msg.to.map((u) => u.to)
                        );
                      }}
                    >
                      Reply All
                    </button>
                    <button onClick={() => onDelete(msg.mid)}>Delete</button>
                  </div>
                ) : (
                  ""
                )}
              </div>
            ))}
      </div>
    </div>
  );
};
