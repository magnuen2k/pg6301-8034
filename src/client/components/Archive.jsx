import React, { useContext } from "react";
import { InboxView } from "./InboxView";
import { UserContext } from "../contexts/context";
import { useLoading } from "../hooks/useLoading";
import { LoadingView } from "./LoadingView";
import { ErrorView } from "./ErrorView";

export const Archive = ({ messageApi }) => {
  const { loading, data, error } = useLoading(() => messageApi.getArchive());
  const { user } = useContext(UserContext);

  if (error) {
    return <ErrorView error={error} />;
  }

  if (!user) {
    return <div className="container">Please log in</div>;
  }

  if (loading) {
    return <LoadingView />;
  }

  if (!data) {
    return <div className="container">No messages</div>;
  }

  return (
    <div className="container">
      <h1>Archived messages</h1>
      <InboxView data={data} />
    </div>
  );
};
