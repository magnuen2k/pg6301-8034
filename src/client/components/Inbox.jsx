import React, { useContext, useEffect, useState } from "react";
import { UserContext } from "../contexts/context";
import { SendMessage } from "./SendMessage";
import { InboxView } from "./InboxView";
import { useHistory } from "react-router-dom";

export const Inbox = ({ messageApi }) => {
  const { user } = useContext(UserContext);
  const [isInbox, setIsInbox] = useState(true);
  const [data, setData] = useState([]);
  const [userMessages, setUserMessages] = useState();
  const history = useHistory();

  useEffect(() => {
    messageApi.getInbox().then((res) => {
      if (res) {
        setData(res);
      }
    });
  }, [isInbox]);

  const toggleView = () => {
    setIsInbox((prevIsInbox) => !prevIsInbox);
  };

  const sendMessage = async (formData) => {
    await messageApi.sendMessage(formData);
    toggleView();
  };

  const reply = async (mid, reply, to) => {
    const formData = {
      message: reply,
      from: user.username,
      replyTo_id: mid,
      to,
    };
    const res = await messageApi.replyToMessage(formData);
    console.log(res);
  };

  const handleSentMessages = async () => {
    const userMessages = await messageApi.getUserMessages();
    setUserMessages(userMessages);
  };

  if (!user) {
    return "log in please";
  }

  if (!data) {
    return "loading";
  }

  return (
    <div className="container">
      <h1>{user.username}'s Inbox</h1>
      <button onClick={toggleView}>
        {!isInbox ? "Return to inbox" : "Send new message"}
      </button>
      {isInbox ? (
        <InboxView data={data} onReply={reply} />
      ) : (
        <SendMessage
          username={user.username}
          onSendMessage={sendMessage}
          messageApi={messageApi}
        />
      )}
      {userMessages && <InboxView data={userMessages} />}
      <div onClick={handleSentMessages}>
        Click here to see your sent messages
      </div>
    </div>
  );
};
