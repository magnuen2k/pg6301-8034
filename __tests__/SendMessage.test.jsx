import React from "react";
import { act, Simulate } from "react-dom/test-utils";
import ReactDOM from "react-dom";
import { UserContext } from "../src/client/contexts/context";
import { MemoryRouter } from "react-router-dom";
import { SendMessage } from "../src/client/components/SendMessage";
async function renderForTest(child, user) {
  const container = document.createElement("div");
  await act(async () => {
    await ReactDOM.render(
      <UserContext.Provider value={{ user }}>
        <MemoryRouter>{child}</MemoryRouter>
      </UserContext.Provider>,
      container
    );
  });
  return container;
}

describe("send message view", () => {
  it("renders send message form", async () => {
    const username = "test";
    const onSendMessage = jest.fn();
    const messageApi = {
      getUsersToMessage: async () => ["test"],
    };
    const container = await renderForTest(
      <SendMessage
        username={username}
        onSendMessage={onSendMessage}
        messageApi={messageApi}
      />
    );
    expect(container.innerHTML).toMatchSnapshot();
  });
  it("tries to send message", async () => {
    const username = "test";
    const onSendMessage = jest.fn();
    const messageApi = {
      getUsersToMessage: async () => ["test"],
    };
    const container = await renderForTest(
      <SendMessage
        username={username}
        onSendMessage={onSendMessage}
        messageApi={messageApi}
      />
    );

    // Add a recipient
    Simulate.change(container.querySelector(".addRecipient input"), {
      target: { value: "test" },
    });
    Simulate.submit(container.querySelector(".addRecipient"));

    // Write a message
    Simulate.change(container.querySelector(".sendMessage textarea "), {
      target: { value: "test message" },
    });

    Simulate.submit(container.querySelector(".sendMessage"));

    expect(onSendMessage).toBeCalled();
  });
});
