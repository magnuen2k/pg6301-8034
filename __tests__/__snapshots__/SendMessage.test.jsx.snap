// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`send message view renders send message form 1`] = `"<div class=\\"sendMessageWrapper\\"><div class=\\"recipients\\">Sender: test</div><div class=\\"recipients\\">Recipients: </div><hr><form class=\\"addRecipient\\"><p>Choose who you want to send a message to</p><label>Pick recipients<select name=\\"recipients\\"><option value=\\"\\" disabled=\\"\\">--------</option><option value=\\"test\\">test</option></select><div><label>Type recipient: <input name=\\"recipient\\"></label></div></label><button class=\\"recipientBtn\\" type=\\"submit\\">+</button><hr></form><form class=\\"sendMessage\\"><label>Your message: <textarea name=\\"message\\"></textarea></label><button type=\\"submit\\">Send message</button></form></div>"`;
